UPGRADE-GUIDE.md

```diff
@@ -0,0 +1,159 @@
+
+This ensures middleware can predictably update headers as needed.
+
+### Response Headers must be lower case
+
+Rack 3 requires all response headers to be lower case. This is to simplify fetching and updating response headers. Previously you had to use something like `Rack::HeadersHash`
+
+```ruby
+def call(env)
+	response = @app.call(env)
+	# HeaderHash must allocate internal objects and compute lower case keys:
+	headers = Rack::Utils::HeaderHash[response[1]]
+
+	cache_response(headers['ETag'], response)
+
+	...
+end
+```
+
+but now you must just use the normal form for HTTP header:
+
+```ruby
+def call(env)
+	response = @app.call(env)
+	# A plain hash with lower case keys:
+	headers = response[1]
+
+	cache_response(headers['etag'], response)
+
+	...
+end
+```
+
+### Multiple response header values are encoded using an `Array`
+
+Response header values can be an Array to handle multiple values 
+(and no longer supports `\n` encoded headers). If you use `Rack::Response`, you don't need to do anything, but if manually append values to response headers, you will need to promote them to an Array, e.g.
+
+```ruby
+def set_cookie_header!(headers, key, value)
+	if header = headers[SET_COOKIE]
+		if header.is_a?(Array)
+			header << set_cookie_header(key, value)
+		else
+			headers[SET_COOKIE] = [header, set_cookie_header(key, value)]
+		end
+	else
+		headers[SET_COOKIE] = set_cookie_header(key, value)
+	end
+end
+```
+
+### Response body might not respond to `#each`
+
+Rack 3 has more strict requirements on response bodies. Previously, response body would only need to respond to `#each` and optionally `#close`. In addition, there was no way to determine whether it was safe to call `#each` and buffer the response.
+
+### Response bodies can be buffered if they expose `#to_ary`
+
+If your body responds to `#to_ary` then it must return an `Array` whose contents are identical to that produced by calling `#each`. If the body responds to both `#to_ary and `#close` then its implementation of `#to_ary` must also call `#close`.
+
+Previously, it was not possible to determine whether a response body was immediately available (could be buffered) or was streaming chunks. This case is now unambiguously exposed by `#to_ary`:
+
+```ruby
+def call(env)
+	status, headers, body = @app.call(env)
+
+	# Check if we can buffer the body into an Array, so we can compute a digest:
+	if body.respond_to?(:to_ary)
+		body = body.to_ary
+		digest = digest_body(body)
+		headers[ETAG_STRING] = %(W/"#{digest}") if digest
+	end
+
+	return [status, headers, body]
+end
+```
+
+### Middleware should not directly modify the response body 
+
+Be aware that the response body might not respond to `#each` and you must now check if the body responds to `#each` or not to determine if it is an enumerable or streaming body.
+
+You must not call `#each` directly on the body and instead you should return a new body that calls `#each` on the original body and yields at least once per iteration.
```

Comment:
I would remove ` and yields at least once per iteration`.  It should be fine to return a body the calls each on the original body, but doesn't yield empty strings, or buffers data to yield in larger chunks instead of smaller chunks.

Reply:
100% agree, updated.